README.md
# membrane.term

[![cljdoc](https://cljdoc.org/badge/com.phronemophobic/membrane.term)](https://cljdoc.org/d/com.phronemophobic/membrane.term)
[![clojurians slack](https://img.shields.io/badge/slack-join_chat-brightgreen.svg)](https://clojurians.slack.com/archives/C02KE09HMHV)
[![clojars](https://img.shields.io/clojars/v/com.phronemophobic/membrane.term.svg)](https://clojars.org/com.phronemophobic/membrane.term)

A simple terminal emulator in clojure. Built with [membrane](https://github.com/phronmophobic/membrane).

## Rationale

I thought it would be fun. There's not much code. All of the ANSI parsing work is done by [asciinema/vt](https://github.com/asciinema/vt). `membrane.term` is just small UI on top.

Some reasons to use **membrane.term**:
- If you need to programmatically control, inspect, drive, or manipulate a terminal
- If you want to learn more about how terminal emulators work
- If you'd like to embed a terminal somewhere. There's not really a guide for embedding, but if you file an issue, I can provide some tips
- If you'd like to capture terminal screenshots for documentation

## Status

Pre 1.0 release.
APIs and behavior are subject to change.
macOS and Linux are the current focus, Windows support is pre-alpha.
## Installation

Membrane.term works on macOS and Linux operating systems.

To invoke via `clj -M:membrane.term` add the following alias to your `deps.edn`:
```Clojure
{:aliases
 {:membrane.term
  {:replace-deps
   {com.phronemophobic/membrane.term {:git/url "https://github.com/phronmophobic/membrane.term.git"
                                      :git/sha "7b973adc4910729042736d8a84e5fce4f4f43722"}
    ;; the next 2 optional deps appease pty4j which does some logging
    ;; change nop to simple if you'd actually like to see the log lines
    org.slf4j/log4j-over-slf4j {:mvn/version "1.7.32"}
    org.slf4j/slf4j-nop {:mvn/version "1.7.32"}}
   :main-opts ["-m" "com.phronemophobic.membrane.term.main"]}}
 :mvn/repos
 ;; pty4j is currently only available from the jetbrains maven repo
 {"jetbrains-intellij-dependencies" {:url "https://packages.jetbrains.team/maven/p/ij/intellij-dependencies"}}}
```

## Command Line Usage

For usage help run:
```bash
clj -M:membrane.term --help
```

### Run an interactive GUI terminal

```bash
clj -M:membrane.term run-term
```

A demo of the types of things you can do:

![run-term-demo](doc/images/run-term-demo.gif?raw=true)

You can also optional specify:
- [`-c, --command`](#command) Command to launch with, defaults to:
  - macOS & Linux: `$SHELL` else `/bin/bash`
  - Windows: `powershell`
- `-w, --width` width of the terminal window in character columns (default: 90)
- `-h, --height` height of the terminal window in character rows (default: 30)
- [`--color-scheme`](#color-schemes) choose a different color scheme
- [`--font-family`](#fonts) choose an OS installed font (default: monospace)
- [`--font-size`](#fonts) specify font point size (default: 12)
- [`--toolkit`](#toolkits) specify the toolkit (default: java2d)

### Run a headless terminal and take a screenshot

Play a script in a headless terminal and write an image to terminal.png.

Given play script `play-msgcat.sh`:
<!-- copied from doc/examples/play-msgcat.sh -->
```bash
# override prompt to something doc-friendly
/bin/bash  # could alternatively use --command
export PS1="$ "
# print out some  colors
clear
msgcat --color=test | head -11
```

If we run:
```
clj -M:membrane.term screenshot --play play-msgcat.sh --height 14
```

Membrane.term passes the script to the terminal character by character, then writes a screenshot to `terminal.png`:

<!-- generated by script/regen-screenshots.sh -->
![play-msgcat-example](doc/images/screenshot-msgcat.png)

You can also optionally specify:
- [`-c, --command`](#command) Command to launch with, defaults to:
  - macOS & Linux: `$SHELL` else `/bin/bash`
  - Windows: `powershell`
- `-p, --play` script to play in the terminal
- `-w, --width` width of the terminal window in character columns (default: 90)
- `-h, --height` height of the terminal window in character columns (default: 30)
- `-o, --out` filename of the image file to generate (default: terminal.png)\
Image format determined by filename extension, supported extensions are `.png`, `.webp`, `.jpeg` and `.jpg`.
- [`--color-scheme`](#color-schemes) choose a different color scheme
- [`--font-family`](#fonts) choose an OS installed font (default: monospace)
- [`--font-size`](#fonts) specify font point size (default: 12)
- [`--toolkit`](#toolkits) specify the toolkit (default: java2d)
- `--line-delay` if your play script runs a command like `lein repl`, the underlying program might not be ready to accept input immediately.\
You can specify a delay in ms to wait after each line is played to the terminal (default: 1000)
- `--final-delay` after the last line of your play script is run, you might need to wait for it to generate all of its output.\
You can specify a delay in ms to wait after the entire script has been played to the terminal  (default: 10000)

Given play script `play-deep-diff.sh`:
<!-- copied from doc/examples/play-deep-diff.sh -->
```bash
# our setup
/bin/bash # could alternatively use --command
export PS1="$ "
cd
mkdir -p target/term-screenshot
cd target/term-screenshot
echo '{:deps {lambdaisland/deep-diff2 {:mvn/version "2.0.108"}}}' > deps.edn
# we clear to wipe setup output
clear
clojure
(require '[lambdaisland.deep-diff2 :as ddiff])
(ddiff/pretty-print (ddiff/diff {:a 1 :b 2} {:a 1 :c 3}))
```
Let's increase the line delay to give the Clojure REPL a chance to start up before feeding it input.
We'll also reduce the final delay, we just don't need 10 seconds for this one:
```bash
clj -M:membrane.term screenshot --play play-deep-diff.sh \
  --width 80 --height 9 \
  --final-delay 1000 --line-delay 3000 \
  --out deep-diff.jpg
```
Produces `deep-diff.jpg`:
<!-- generated by script/regen-screenshots.sh -->
![play-deep-diff-example](doc/images/screenshot-deep-diff.png)


## Command

By default, `membrane.term` will attempt to launch with your OS shell.

Use the `--command` option to launch with a different program.

Let's suppose your favorite process monitor is [htop](https://htop.dev/) and that just want to run that:
```bash
clj -M:membrane.term run-term --command "htop --tree --delay 10"
```

Sometimes you are interested in taking a screenshot of only the output of a command:
```bash
clj -M:membrane.term screenshot --command "msgcat --color=test" --width 80 --height 21
```
![screenshot-msgcat-command-example](doc/images/screenshot-msgcat-command.png)

Or, for a contrived example, let's pretend you want to play some text in `vi` and take a screenshot.
Given file `an-ode-to-clojure.txt`:
```
iRoses are red
Violets are blue
Clojure is fun
And so are you
```

Let's run:
```bash
clj -M:membrane.term screenshot --command vi --play an-ode-to-clojure.txt \
  --width 40 --height 10
```
![screenshot-vi-command-example](doc/images/screenshot-vi-command.png)

## Color Schemes

There are boatloads of terminal color schemes available at [iTerm2-Color-Schemes](https://github.com/mbadolato/iTerm2-Color-Schemes).
Review what you might like [from their screenshots](https://github.com/mbadolato/iTerm2-Color-Schemes#screenshots), then find the corresponding [`.itermcolors` file under their schemes directory](https://github.com/mbadolato/iTerm2-Color-Schemes/tree/master/schemes).

Use the `--color-scheme` option to specify an `.itermcolors` file, either from a copy you have downloaded:

```bash
clj -M:membrane.term run-term --width 90 --height 30 \
  ---color-scheme "Builtin Solarized Dark.itermcolors"
```

...or directly from the raw GitHub URL:

```bash
clj -M:membrane.term screenshot --play play-msgcat.sh --height 14 \
  --color-scheme "https://raw.githubusercontent.com/mbadolato/iTerm2-Color-Schemes/master/schemes/Builtin%20Solarized%20Dark.itermcolors"
```
The screenshot command produces `terminal.png`:
<!-- generated by script/regen-screenshots.sh -->
![play-msgcat-scheme-example](doc/images/screenshot-msgcat-scheme.png)

## Fonts

You can specify a font family and size.
The font must be OS installed and is assumed to be monospace.

Let's pretend that, for whatever reason, you've fallen in love with the [Nova Mono font](https://fonts.google.com/specimen/Nova+Mono), and have installed it on your system.

To use this font in an interactive membrane.term terminal, specify its installed name. Here we've opted for a point size of 16:
```bash
clj -M:membrane.term run-term --font-family "NovaMono" --font-size 16
```

The same options are available for screenshots:
```bash
clj -M:membrane.term screenshot --play play-msgcat.sh --height 14 \
  --font-family "NovaMono" --font-size 16
```

produces `terminal.png`:
<!-- generated by script/regen-screenshots.sh -->
![play-msgcat-font-example](doc/images/screenshot-msgcat-font.png)

## Toolkits

You can specify a toolkit. Available options are "java2d" (the default) and "skia".

### Java2d

The java2d toolkit will use swing for windowing, graphics, and events.

### Skia

The skia toolkit will use [skia](https://skia.org/) for graphics and [glfw](https://www.glfw.org/) for windowing and events. Using the skia toolkit requires the following extra dependency:

M1 mac:
```clojure
com.phronemophobic.membrane/skialib-macosx-aarch64 {:mvn/version "0.9.31.0-beta"}
```

Non-M1 mac:
```clojure
com.phronemophobic.membrane/skialib-macosx-x86-64 {:mvn/version "0.9.31.0-beta"}
```

Linux
```clojure
com.phronemophobic.membrane/skialib-linux-x86-64 {:mvn/version "0.9.31.0-beta"}
```

If running from the `membrane.term` project, you can use the `:skia` alias to add the dependency:

```bash
clojure -M:membrane.term:skia run-term --toolkit skia
```

## Discuss

Questions? Comments? Connect with us on clojurians slack in [#membrane-term](https://clojurians.slack.com/archives/C02KE09HMHV) (join [here](http://clojurians.net/))

## License

Copyright © 2021 Adrian Smith

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Can now specify the launch command [#3](https://github.com/phronmophobic/membrane.term/issues/3)
- Document public API [#28](https://github.com/phronmophobic/membrane.term/issues/28)
- **Breaking** `screenshot` `:play` option was file, is now string (command line is unaffected, `--play` still references a file) [#26](https://github.com/phronmophobic/membrane.term/issues/28)

## [0.9.0] - 2021-11-30
- Initial release to clojars


script/ansi-text-test.clj
#!/usr/bin/env bb

(ns ansi-text-test
  "A little script to print out some sgr code decorated text for manual visual review.
   An attempt was made to fit entire output on a reasonably sized screen."

  (:require [clojure.set :as cset]
            [clojure.string :as string]))

;; from https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters
(def sgr-codes {:clear 0
                :bold 1
                :dim 2
                :italic 3
                :underline 4
                :slow-blink 5
                :rapid-blink 6
                :inverted 7
                :hide 8
                :strikeout 9
                :primary-font 10
                :alt-font-1 11
                :alt-font-2 12
                :alt-font-3 13
                :alt-font-4 14
                :alt-font-5 15
                :alt-font-6 16
                :alt-font-7 17
                :alt-font-8 18
                :alt-font-9 19
                :fraktur 20
                :not-bold 21 ;; or double underline sometimes
                :normal-intensity 22
                :not-italic 23
                :not-underlined 24
                :not-blinking 25
                :proportional 26
                :not-inverted 27
                :reveal 28
                :not-strikeout 29
                :fg-black 30
                :fg-red 31
                :fg-green 32
                :fg-yellow 33
                :fg-blue 34
                :fg-magenta 35
                :fg-cyan 36
                :fg-white 37
                :fg-rgb 38
                :fg-default 39
                :bg-black 40
                :bg-red 41
                :bg-green 42
                :bg-yellow 43
                :bg-blue 44
                :bg-magenta 45
                :bg-cyan 46
                :bg-white 47
                :bg-rgb 48
                :bg-default 49

                :disable-proportional 50

                :framed 51
                :encircled 52
                :overlined 53
                :not-framed-nor-circled 54
                :not-overlined 55

                :underline-color 58
                :default-underline-color 59

                :ideogram-underline 60
                :ideogram-double-underline 61
                :ideogram-overline 62
                :ideogram-double-overline 63
                :ideogram-stress 64
                :ideogram-clear 65

                :superscript 73
                :subscript 74
                :not-super-sub-script 75

                :fg-grey 90
                :fg-bright-red 91
                :fg-bright-green 92
                :fg-bright-yellow 93
                :fg-bright-blue 94
                :fg-bright-magenta 95
                :fg-bright-cyan 96
                :fg-bright-white 97

                :bg-grey 100
                :bg-bright-red 101
                :bg-bright-green 102
                :bg-bright-yellow 103
                :bg-bright-blue 104
                :bg-bright-magenta 105
                :bg-bright-cyan 106
                :bg-bright-white 107})


(def keys-visited (atom #{}))

(def isgr-codes (cset/map-invert sgr-codes))

;; to protect myself from myself accces maps from these fns only:
(defn- get-sgr-code [k]
  (let [code (or (get sgr-codes k) (throw (ex-info (format "oops key %s not defined" k) {})))]
    (swap! keys-visited conj k)
    code))
(defn- get-sgr-key [code]
  (let [k (or (get isgr-codes code) (throw (ex-info (format "oops code %s not defined" code) {})))]
    (swap! keys-visited conj k)
    k))

(defn- ansi [code & params]
  (str "\u001b[" (string/join ";" (concat [code] params)) "m"))

(defn- ansi-by-key [k & params]
  (apply ansi (get-sgr-code k) params))

(defn- sample-text [k & params]
  (let [lbl (name k)
        code (get-sgr-code k)]
    (format "%s%03d%s-%s "
            (apply ansi code params)
            code
            (if (seq params)
              (str ";" (string/join ";" params))
              "")
            lbl)))

(defn sample-text-for-codes [codes]
  (map (fn [c]
         (let [k (get-sgr-key c)]
           (str (sample-text k) (ansi-by-key :clear))))
       codes))

(defn sample-text-for-keys [ks]
  (sample-text-for-codes
    (map get-sgr-code ks)))

(defn- stripped [s]
  (string/replace s #"\u001b\[.*?m" ""))

(defn print-samples [title samples]
  (print (format "\n> %s\n  " title))
  (let [max-len (->> samples
                     (map stripped)
                     (map count)
                     (reduce max))
        col-width (+ max-len 4)
        screen-width 120
        num-cols (if (> col-width screen-width) 1 (int (/ screen-width col-width)))
        col-width (/ screen-width num-cols)]
    (->> samples
         (map (fn [s] (let [len (count (stripped s))
                            pad (max 0 (- col-width len 2))]
                        (str "[" s
                             (apply str (repeat pad " "))
                             "]"))))
         (partition num-cols num-cols nil)
         (interpose ["\n  "])
         (apply concat)
         (run! print)))
  (println))

(print-samples "membrane.term - supported"
               (sample-text-for-keys
                [:clear
                 :bold
                 :italic
                 :normal-intensity
                 :not-italic
                 :fg-black
                 :fg-red
                 :fg-green
                 :fg-yellow
                 :fg-blue
                 :fg-magenta
                 :fg-cyan
                 :fg-white
                 :fg-rgb
                 :fg-default
                 :bg-black
                 :bg-red
                 :bg-green
                 :bg-yellow
                 :bg-blue
                 :bg-magenta
                 :bg-cyan
                 :bg-white
                 :bg-rgb
                 :bg-default
                 :fg-grey
                 :fg-bright-red
                 :fg-bright-green
                 :fg-bright-yellow
                 :fg-bright-blue
                 :fg-bright-magenta
                 :fg-bright-cyan
                 :fg-bright-white
                 :bg-grey
                 :bg-bright-red
                 :bg-bright-green
                 :bg-bright-yellow
                 :bg-bright-blue
                 :bg-bright-magenta
                 :bg-bright-cyan
                 :bg-bright-white]))

(print-samples "membrane.term - questionable, 21 treated as not-bold, iTerm2 ignores, GNOME double underlines"
               [(str (sample-text :bold)
                     (sample-text :not-bold)
                     (ansi-by-key :clear))])

(print-samples "membrane.term - interesting candidates"
               (sample-text-for-keys
                [:dim
                 :underline
                 :slow-blink
                 :inverted
                 :strikeout
                 :not-underlined
                 :not-blinking
                 :not-inverted
                 :not-strikeout]))

(print-samples "membrane.term - maybe someday?"
               (sample-text-for-keys
                [:rapid-blink
                 :hide
                 :reveal
                 :overlined
                 :not-overlined
                 :underline-color
                 :default-underline-color]))

(print-samples "sanity on/off tests"
               [(str (sample-text :bold)
                     (sample-text :dim)
                     (sample-text :normal-intensity)
                     (ansi-by-key :clear))
                (str (sample-text :italic)
                     (sample-text :not-italic)
                     (ansi-by-key :clear))
                (str (sample-text :underline)
                     (sample-text :not-underlined)
                     (ansi-by-key :clear))
                (str (sample-text :slow-blink)
                     (sample-text :not-blinking)
                     (ansi-by-key :clear))
                (str (sample-text :fg-red)
                     (sample-text :inverted)
                     (ansi-by-key :clear))
                (str (sample-text :inverted)
                     (sample-text :not-inverted)
                     (ansi-by-key :clear))
                (str (sample-text :strikeout)
                     (sample-text :not-strikeout)
                     (ansi-by-key :clear))])

(print-samples "sanity rgb tests"
               [(str (sample-text :bg-rgb 5 130)
                     (ansi-by-key :clear))
                (str (sample-text :bg-rgb 2 100 200 43)
                     (ansi-by-key :clear))
                (str (sample-text :fg-rgb 5 33)
                     (sample-text :bg-rgb 5 192)
                     (sample-text :bg-default)
                     (ansi-by-key :clear))])

(print-samples "interesting, these work for GNOME Terminal"
               [(str (sample-text :rapid-blink)
                     (sample-text :not-blinking)
                     (ansi-by-key :clear))
                (str (sample-text :overlined)
                     (sample-text :not-overlined)
                     (ansi-by-key :clear))
                (str (sample-text :underline)
                     (sample-text :underline-color 5 130)
                     (sample-text :underline-color 2 52 180 235)
                     (sample-text :default-underline-color)
                     (ansi-by-key :clear))])

(print-samples "who supports hide/reveal? GNOME Terminal!"
               [(str (sample-text :hide)
                     (sample-text :reveal)
                     (ansi-by-key :clear))])

(print-samples "codes expected to be no-op (update as we find current terminals that support in the wild)"
               (sample-text-for-keys
                [:primary-font
                 :alt-font-1
                 :alt-font-2
                 :alt-font-3
                 :alt-font-4
                 :alt-font-5
                 :alt-font-6
                 :alt-font-7
                 :alt-font-8
                 :alt-font-9
                 :fraktur
                 :proportional
                 :disable-proportional
                 :framed
                 :encircled
                 :not-framed-nor-circled
                 :ideogram-underline
                 :ideogram-double-underline
                 :ideogram-overline
                 :ideogram-double-overline
                 :ideogram-stress
                 :ideogram-clear
                 :superscript
                 :subscript
                 :not-super-sub-script]))

(let [missed (apply dissoc sgr-codes @keys-visited)]
  (if (seq missed)
    (println "\n* Heya, your calls didn't hit: " missed)
    (println "\nAll defined sgr codes hit.")))


test/com/phronemophobic/membrane/term_test.clj
(ns com.phronemophobic.membrane.term-test
  (:require [clojure.test :refer :all]
            [com.phronemophobic.membrane.term :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


script/regen-screenshots.sh
#!/usr/bin/env bash

set -eou pipefail

echo "Regenerating screenshots for README"

echo "- msgcat -"
clj -M:membrane.term screenshot \
    --play doc/examples/play-msgcat.sh \
    --height 14 \
    --out doc/images/screenshot-msgcat.png

echo "- deep-diff -"
clj -M:membrane.term screenshot \
    --play doc/examples/play-deep-diff.sh \
    --width 80 --height 9 \
    --final-delay 1000 --line-delay 3000 \
    --out doc/images/screenshot-deep-diff.png

echo "- msgcat with color scheme -"
clj -M:membrane.term screenshot \
    --play doc/examples/play-msgcat.sh \
    --height 14 \
    --color-scheme "doc/examples/Builtin Solarized Dark.itermcolors" \
    --out doc/images/screenshot-msgcat-scheme.png

echo "- msgcat with font -"
clj -M:membrane.term screenshot \
    --play doc/examples/play-msgcat.sh \
    --height 14 \
    --font-family "NovaMono" \
    --font-size 16 \
    --out doc/images/screenshot-msgcat-font.png

echo "- launch msgcat as command -"
clj -M:membrane.term screenshot \
    --command 'msgcat --color=test' \
    --width 80 --height 22 \
    --out doc/images/screenshot-msgcat-command.png

echo "- launch with vi and play some text -"
clj -M:membrane.term screenshot \
    --command vi --play doc/examples/an-ode-to-clojure.txt \
    --width 40 --height 10 \
    --out doc/images/screenshot-vi-command.png


src/com/phronemophobic/membrane/term/color_scheme.clj
(ns ^:no-doc com.phronemophobic.membrane.term.color-scheme
  (:require [clojure.data.xml :as xml]
            [clojure.data.zip.xml :as zxml]
            [clojure.set :as cset]
            [clojure.java.io :as io]
            [clojure.zip :as zip]))

(def ^:private iterm-color
  ;; iterm keys currently of interest to us, others can be added if/when we add support in membrane.term
  {"Ansi 0 Color"        :black
   "Ansi 1 Color"        :red
   "Ansi 2 Color"        :green
   "Ansi 3 Color"        :yellow
   "Ansi 4 Color"        :blue
   "Ansi 5 Color"        :magenta
   "Ansi 6 Color"        :cyan
   "Ansi 7 Color"        :white
   "Ansi 8 Color"        :bright-black
   "Ansi 9 Color"        :bright-red
   "Ansi 10 Color"       :bright-green
   "Ansi 11 Color"       :bright-yellow
   "Ansi 12 Color"       :bright-blue
   "Ansi 13 Color"       :bright-magenta
   "Ansi 14 Color"       :bright-cyan
   "Ansi 15 Color"       :bright-white
   "Cursor Color"        :cursor
   "Cursor Text Color"   :cursor-text
   "Background Color"    :background
   "Foreground Color"    :foreground})

(def ^:private required-colors (cset/map-invert iterm-color))

(def ^:private iterm-component
  {"Red Component" :red
   "Green Component" :green
   "Blue Component" :blue
   "Alpha Component" :alpha})

(def ^:private required-components (select-keys (cset/map-invert iterm-component)
                                                [:red :green :blue]))

(defn- iterm-color->rgb [ctx zdict]
  (let [cmap (->> (iterate #(-> % zip/right zip/right) (zip/down zdict))
                  (take-while identity)
                  (reduce (fn [acc z]
                            (let [cin  (-> z zxml/text)]
                              (if-let [ckey (get iterm-component cin)]
                                (assoc acc ckey (-> z zip/right zxml/text Double/parseDouble))
                                acc)))
                          {}))
        missing-components (apply dissoc required-components (keys cmap))]
    (if (seq missing-components)
      (throw (ex-info (format "%s %s is missing components: %s" (:source ctx) (pr-str (:context ctx)) (vec (vals missing-components))) {}))
      (if (:alpha cmap)
        ((juxt :red :green :blue :alpha) cmap)
        ((juxt :red :green :blue) cmap)))))

(defn load-scheme
  "We currently support iTerm scheme format which is held in Apple Info.plist XML.

  `source` can be whatever `clojure.java.io/reader` accepts, common usages:
  - local path `\"./Solarized Dark.itermcolors\"` or
  - a url `\"https://raw.githubusercontent.com/mbadolato/iTerm2-Color-Schemes/master/schemes/Dracula%2B.itermcolors\"`"
  [source]
  (with-open [rdr (io/reader source)]
    (let [z (-> rdr
                (xml/parse :namespace-aware false :skip-whitespace true)
                zip/xml-zip)
          scheme (reduce (fn [acc z]
                           (let [cin (-> z zxml/text)]
                             (if-let [color-name (get iterm-color cin)]
                               (let [rgba (->> z zip/right (iterm-color->rgb {:context cin :source source}))]
                                 (assoc acc color-name rgba))
                               acc)))
                         {}
                         (->> (iterate #(-> % zip/right zip/right) (zxml/xml1-> z :dict :key))
                              (take-while identity)))
          missing-keys (apply dissoc required-colors (keys scheme))]
      (if (seq missing-keys)
        (throw (ex-info (format "%s is missing colors: %s" source (vec (vals missing-keys))) {}))
        scheme))))


src/com/phronemophobic/membrane/term/main.clj
(ns ^:no-doc com.phronemophobic.membrane.term.main
  (:require [clojure.java.io :as io]
            [clojure.string :as string]
            [com.phronemophobic.membrane.term :as term]
            [membrane.ui :as ui]
            [membrane.toolkit :as tk]
            [com.phronemophobic.membrane.term.color-scheme :as color-scheme]
            [docopt.core :as docopt]))

(def docopt-usage
  "membrane.term

Usage:
  membrane.term run-term [--width=<cols>] [--height=<rows>] \\
   [--color-scheme=<path>] [--font-family=<font>] [--font-size=<points>] [--toolkit=<toolkit>]
  membrane.term screenshot --play=<path> [--width=<cols>] [--height=<rows>] \\
   [--color-scheme=<path>] [--font-family=<font>] [--font-size=<points>]\\
   [--out=<file>] [--line-delay=<ms>] [--final-delay=<ms>]  [--toolkit=<toolkit>]
  membrane.term --help

Common Options:
  -w, --width=<cols>         Width in characters [default: 90]
  -h, --height=<rows>        Height in characters [default: 30]
      --color-scheme=<path>  Local path or url to iTerm .itermcolors scheme file, uses internal scheme by default.
      --font-family=<font>   Choose an OS installed font [default: monospace]
      --font-size=<points>   Specify the font point size [default: 12]
      --toolkit=<toolkit>    Specify the toolkit [default: java2d]

Screenshot Options:
  -p, --play=<path>          Path to script to play in terminal
  -o, --out=<file>           Filename of the image to generate [default: terminal.png]
      --line-delay=<ms>      Delay in ms to wait after each line is sent to the terminal [default: 1000]
      --final-delay=<ms>     Delay in ms to wait before writing the view of the terminal [default: 10000]

Replace membrane.term with your appropriate Clojure tools CLI launch sequence. For example:
|
| clojure -M:membrane.term run-term -w 133 -h 60
|")

(defn parse-font-family [v]
  (if (= "monospace" v)
    :monospace
    (str v)))

(defn- int-parser-validator [min]
  (fn [v]
    (let [num (if (integer? v)
                v
                (when (and (string? v) (re-matches #"\d+" v))
                  (Integer/parseInt v)))]
      (if (and num (>= num min))
        num
        {:error (format "expected positive integer >= %d" min)}))))

(defn- parse-existing-path [v]
  (when v
    (let [p (str v)]
      (try
        (with-open [_rdr (io/reader p)])
        p
        (catch Throwable e
          {:error (format "unable to open path, %s" (ex-message e))})))))

(defn- parse-play-script [v]
  (when v
    (let [p (parse-existing-path v)]
      (if (:error p)
        p
        (slurp p)))))

(defn- parse-color-scheme [v]
  (when v
    (let [p (parse-existing-path v)]
      (if (:error p)
        p
        (try
          (color-scheme/load-scheme p)
          (catch Throwable e
            {:error (format "Unable to load color-scheme from %s" p)
             :exception e}))))))

(defn- parse-image-out [v]
  (when v
    (let [p (str v)]
      (if (re-find #"(?i).+\.(png|jpeg|jpg|webp)$" p)
        p
        {:error "supported image formats are png, webp and jpeg (aka jpg)."}))))


(def valid-toolkits #{"java2d" "skia"})
(defn- load-toolkit [toolkit]
  (if-not (or (nil? toolkit)
              (valid-toolkits toolkit))
    (throw (ex-info (format "Invalid toolkit: %s. Valid toolkits are %s."
                            toolkit
                            (->> valid-toolkits
                                 (map #(str "\"" % "\""))
                                 (string/join ", ")))
                    {}))
    (case toolkit
      (nil "java2d")
      @(requiring-resolve 'membrane.java2d/toolkit)

      ("skia")
      @(requiring-resolve 'membrane.skia/toolkit))))

(defn parse-toolkit [v]
  (try
    (load-toolkit v)
    (catch Throwable e
      {:error (ex-message e)})))

(defn- validate-font [args]
  (let [font-family (get args "--font-family")
        font-size (get args "--font-size")
        toolkit (get args "--toolkit")
        font-family (if (keyword? font-family)
                      (tk/logical-font->font-family toolkit :monospace)
                      font-family)]
    (when-not (tk/font-exists? toolkit (ui/font font-family font-size))
      {:error (format "font family %s, size %d not found" font-family font-size)})))

(defn- validate-args [args args-def post-validations]
  (let [parsed-args (reduce-kv (fn [m k v]
                                 (let [od (get args-def k)
                                       v ((or od identity) v)]
                                   (if (:error v)
                                     (reduced (assoc v :error (format "%s: %s" k (:error v))))
                                     (assoc m k v))))
                               {}
                               args)]
    (if (:error parsed-args)
      parsed-args
      (if-let [validation-error (first (keep #(% parsed-args) post-validations))]
        validation-error
        parsed-args))))

(defn- keywordize [arg-map]
  (reduce-kv (fn [m k v]
               (if v
                 (assoc m (keyword (string/replace-first k #"^--" "")) v)
                 m))
             {}
             arg-map))

(defn- undo-line-continuations
  "Docopt does not seem to support line continuations, but I feel they make the usage help readable."
  [usage]
  (string/replace usage #"\\\R" ""))

(defn -main [& args]
  (docopt/docopt (undo-line-continuations docopt-usage) args
                 (fn result-fn [arg-map]
                   (let [arg-map (validate-args arg-map {"--width" (int-parser-validator 1)
                                                         "--height" (int-parser-validator 1)
                                                         "--play" parse-play-script
                                                         "--color-scheme" parse-color-scheme
                                                         "--out" parse-image-out
                                                         "--line-delay" (int-parser-validator 0)
                                                         "--final-delay" (int-parser-validator 0)
                                                         "--font-family" parse-font-family
                                                         "--font-size" (int-parser-validator 1)
                                                         "--toolkit" parse-toolkit}
                                                [validate-font])]
                     (if-let [error (:error arg-map)]
                       (do
                         (println (format "*\n* Error: %s\n*" error))
                         (when-let [e (:exception arg-map)]
                           (println (format "* Exception: %s\n*" (ex-message e))))
                         (println (str "\n" docopt-usage))
                         (System/exit 1))
                       (let [opts (keywordize arg-map)
                             opts (update opts :font-family
                                          (fn [font-family]
                                            (if (keyword? font-family)
                                              (tk/logical-font->font-family (:toolkit opts) font-family)
                                              font-family)))]
                         (cond
                           (:help opts) (println docopt-usage)
                           (:run-term opts) (term/run-term opts)
                           :else (term/screenshot opts))))))
                 (fn usage-fn [_]
                   (println "*\n* Usage error\n*\n")
                   (println docopt-usage)
                   (System/exit 1)))
  (shutdown-agents))
